### PR TITLE
CloudFormation Macro support for two Cloudwatch log group subscription filters

### PR DESCRIPTION
**Pull request by zARODz11z**
### What does this PR do?
This PR edits the logic in the forwarder.ts file to support two Cloudwatch log group subscription filters. Additionally this PR tests the new logic with unit tests in the forwarder.spec.ts file. This PR allows the Datadog Cloudformation macro to take full advantage of AWS cloudwatch log group's support for 2 log group subscription filters.

### Motivation
AWS Cloudwatch log groups added support for up to 2 subscription filters per log group and the Datadog Cloudformation macro logic currently only supports 1 subscription filter per log group.

### Testing Guidelines
I added unit tests for the new cases introduced by Cloudformations support of up to 2 log group subscription filters per log group and additionally manually tested these cases by repeatedly deploying a test SAM Lambda function with the test scenarios and confirming the expected result in the AWS console.

### Types of changes

- [x] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] Misc (docs, refactoring, dependency upgrade, etc.)

### Check all that apply

- [x] This PR's description is comprehensive
- [ ] This PR contains breaking changes that are documented in the description
- [ ] This PR introduces new APIs or parameters that are documented and unlikely to change in the foreseeable future
- [ ] This PR impacts documentation, and it has been updated (or a ticket has been logged)
- [x] This PR's changes are covered by the automated tests
- [ ] This PR collects user input/sensitive content into Datadog

### CODE REVIEW

File: serverless/src/forwarder.ts [hunk 1]
```diff
@@ -8,8 +8,9 @@ const LAMBDA_LOG_GROUP_PREFIX = "/aws/lambda/";
 const FN_SUB = "Fn::Sub";
 const FN_JOIN = "Fn::Join";
 const REF = "Ref";
-export const SUBSCRIPTION_FILTER_NAME = "datadog-serverless-macro-filter";
+const maxAllowableLogGroupSubscriptions = 2;
```
Comment (nhinsch) [inline serverless/src/forwarder.ts:12]: Nit: What's this blank line doing here? I would move it below all of the constants.

Comment (nhinsch) [reply serverless/src/forwarder.ts:12]: Also, looks like the convention in this file/project is for constants like this to be in `ALL_CAPS`

File: serverless/test/forwarder.spec.ts [hunk 2]
```diff
@@ -353,10 +358,75 @@ describe("shouldSubscribeLogGroup", () => {
       [logGroupName]: { logGroup: { logGroupName } },
     });
     const shouldSub = await shouldSubscribeLogGroup(cloudWatchLogs as any, logGroupName);
-    expect(shouldSub).toBeTruthy();
+    expect(shouldSub).toBe(true);
```
Comment (nhinsch) [inline serverless/test/forwarder.spec.ts:361]: 👌

File: serverless/src/forwarder.ts [hunk 3]
```diff
@@ -55,8 +56,10 @@ export class MissingSubDeclarationError extends Error {
  * log group for each lambda function.
  *
  * We first check if a log group exists for a given lambda. If it does, we then check if there
- * are existing subscriptions. We can go ahead and add the subscription to the forwarder ARN
- * (using AWS SDK) if no other unknown subscriptions exist.
+ * are existing subscriptions. We cannot add the subscription to the forwarder ARN if the number
+ * of existing subscriptions equals the maximum number allowed by AWS cloudwatch logs.
+ * We can go ahead and add the subscription to the forwarder ARN
+ * (using AWS SDK) in any other case.
```
Comment (nhinsch) [inline serverless/src/forwarder.ts:62]: What about the case where there is only one subscription, but it is ours? Looks like this comment doesn't match the logic in the code

File: serverless/src/forwarder.ts [hunk 4]
```diff
@@ -113,7 +116,8 @@ export async function addCloudWatchForwarderSubscriptions(
     }
 
     // If log group exists, we need to check if there are any existing subsciption filters.
-    // We will only add a new subscription to the provided forwarder ARN if no current subscriptions exist.
+    // If the number of existing subscription filters equals the maximum allowed by AWS cloudwatch log groups
```
Comment (nhinsch) [inline serverless/src/forwarder.ts:119]: Nit: `cloudwatch` -> `Cloudwatch`

File: serverless/src/forwarder.ts [hunk 5]
```diff
@@ -206,9 +210,30 @@ export async function getExistingLambdaLogGroupsOnStack(cloudWatchLogs: CloudWat
...
+Grab all the subscription filters from the log group
+Loop through the subscription filters, checking if any of the filters is a existing datadog-cloudformation-macro filter
+If we did not find a datadog-cloudformation-macro filter and the number of active subscription filters on the log group 
+is less than the maxAllowableLogGroupSubscriptions from AWS cloudwatch log groups then we subscribe.
+Otherwise we do not subscribe.
+*/
 export async function shouldSubscribeLogGroup(cloudWatchLogs: CloudWatchLogs, logGroupName: string) {
   const subscriptionFilters = await describeSubscriptionFilters(cloudWatchLogs, logGroupName);
-  return subscriptionFilters.length === 0;
+  let foundDatadogSubscriptionFilter = false;
```
Comment (nhinsch) [inline serverless/src/forwarder.ts:222]: I think you could simplify this logic a bit:

- If there are too many filters, return false
- Iterate through the filters, if there is a Datadog one, immediately return false
- If you get this far, then you can return true

Comment (nhinsch) [reply serverless/src/forwarder.ts:222]: The good news is that with your unit tests, it's easy to update the code and make sure that it still works the same way.

File: serverless/src/forwarder.ts [hunk 6]
```diff
@@ -206,9 +210,30 @@ export async function getExistingLambdaLogGroupsOnStack(cloudWatchLogs: CloudWat
   return logGroups ?? [];
 }
 
+/*
+Grab all the subscription filters from the log group
+Loop through the subscription filters, checking if any of the filters is a existing datadog-cloudformation-macro filter
+If we did not find a datadog-cloudformation-macro filter and the number of active subscription filters on the log group 
+is less than the maxAllowableLogGroupSubscriptions from AWS cloudwatch log groups then we subscribe.
+Otherwise we do not subscribe.
+*/
```
Comment (nhinsch) [inline serverless/src/forwarder.ts:219]: I would try to remove this comment -- generally I would only put a comment in code if there is something very non-obvious that the code itself doesn't explain. And even then, a brief 1-liner comment is usually better than a longer one. See if you can make your code so easy to read that no comment is needed!

File: serverless/test/forwarder.spec.ts [hunk 7]
```diff
@@ -137,8 +137,8 @@ describe("addCloudWatchForwarderSubscriptions", () => {
     });
     expect(resources).toEqual(mockResources([lambda])); // template should not be modified
   });
-
-  it("does not overwrite existing unknown subscription on log group", async () => {
+  //At the moment this test assumes a log group subscriptions is full when it has 2 existing subscriptions.
+  it("does not overwrite existing subscriptions on a log group whose subscriptions slots are full", async () => {
```
Comment (nhinsch) [inline serverless/test/forwarder.spec.ts:141]: I would rephrase `a log group whose subscriptions slots are full` -> `a log group that already has the maximum number of subscriptions`. I know it's helpful to think about "slots" or "being full" here but those aren't really terms AWS uses.

File: serverless/test/forwarder.spec.ts [hunk 8]
```diff
@@ -353,10 +358,75 @@ describe("shouldSubscribeLogGroup", () => {
       [logGroupName]: { logGroup: { logGroupName } },
     });
     const shouldSub = await shouldSubscribeLogGroup(cloudWatchLogs as any, logGroupName);
-    expect(shouldSub).toBeTruthy();
+    expect(shouldSub).toBe(true);
   });
-
-  it("returns false if there is an existing subscription", async () => {
+  it("returns false if the log group only has 1 existing datadog-serverless-macro-filter subscription filter.", async () => {
```
Comment (nhinsch) [inline serverless/test/forwarder.spec.ts:363]: I would put a blank line in between each test

Comment (nhinsch) [inline serverless/test/forwarder.spec.ts:363]: It looks like most of the other tests in this file don't end their descriptions with a `.` -- I would remove yours to keep it consistent

File: serverless/src/forwarder.ts (hunk 5 above)
Comment (nhinsch) [reply serverless/src/forwarder.ts:222]: A way to think about this is that you usually want to return an answer as early as possible if you can.

File: serverless/src/forwarder.ts (hunk 6 above)
Comment (nhinsch) [reply serverless/src/forwarder.ts:219]: Your tests will also help document the behavior of the function. So if someone doesn't understand your code, they can look at the unit tests and see the different cases.

File: serverless/src/forwarder.ts (hunk 1 above)
Comment (zARODz11z) [reply serverless/src/forwarder.ts:12]: Do you mean the blank line on line 12?

Comment (zARODz11z) [reply serverless/src/forwarder.ts:12]: I got rid of the space on line 12 but added one on line 13 to separate the interface definitions from the interface definitions as that was the convention when I cloned the repo.

File: serverless/src/forwarder.ts (hunk 3 above)
Comment (zARODz11z) [reply serverless/src/forwarder.ts:62]: I changed that comment to 
* We first check if a log group exists for a given lambda. If it does, we then check if there
 * are existing subscriptions. If we do not find an existing datadog-cloudformation-macro subscription and
 * the number of existing subscription filters is less than the MAX_ALLOWABLE_LOG_GROUP_SUBSCRIPTIONS then 
 * we will go ahead and add the subscription to the forwarder ARN (using AWS SDK). Otherwise we do not add 
 * the subscription.

File: serverless/src/forwarder.ts [hunk 9]
```diff
@@ -208,7 +209,19 @@ export async function getExistingLambdaLogGroupsOnStack(cloudWatchLogs: CloudWat
 
 export async function shouldSubscribeLogGroup(cloudWatchLogs: CloudWatchLogs, logGroupName: string) {
   const subscriptionFilters = await describeSubscriptionFilters(cloudWatchLogs, logGroupName);
-  return subscriptionFilters.length === 0;
+  let foundDatadogSubscriptionFilter = false;
+  const numberOfActiveSubscriptionFilters = subscriptionFilters.length;
+  if (numberOfActiveSubscriptionFilters === MAX_ALLOWABLE_LOG_GROUP_SUBSCRIPTIONS) {
```
Comment (nhinsch) [inline serverless/src/forwarder.ts:214]: What if the number of subscription filters is greater than the number of maximum allowable log group subscriptions? Could happen in the future if AWS changes the maximum on their side.

Comment (zARODz11z) [reply serverless/src/forwarder.ts:214]: Hmmm yeah that could happen if AWS updates and if Datadog does not update the macro, good catch, thanks

File: serverless/src/forwarder.ts [hunk 10]
```diff
@@ -208,7 +209,18 @@ export async function getExistingLambdaLogGroupsOnStack(cloudWatchLogs: CloudWat
...
   const subscriptionFilters = await describeSubscriptionFilters(cloudWatchLogs, logGroupName);
-  return subscriptionFilters.length === 0;
+  const numberOfActiveSubscriptionFilters = subscriptionFilters.length;
+  if (numberOfActiveSubscriptionFilters >= MAX_ALLOWABLE_LOG_GROUP_SUBSCRIPTIONS) {
+    return false;
+  }
+  for (const subscription of subscriptionFilters) {
+    const filterName = subscription.filterName;
+    if (filterName === SUBSCRIPTION_FILTER_NAME) {
+      //We found a existing datadog-cloudformation-macro subscription
```
Comment (nhinsch) [inline serverless/src/forwarder.ts:219]: a existing -> an existing

**Review by nhinsch — APPROVED** [review]
